File: leet/70e_ClimbingStairs.java
public class Solution {
    int[] dp;
    
    public int util(int n) {
        if(dp[n] != -1) return dp[n];
        
        dp[n] = util(n-1) + util(n-2);
        
        return dp[n];
    }

    public int climbStairs(int n) {
        dp = new int[n+1];
        Arrays.fill(dp, -1);
        
        dp[0] = 1; dp[1] = 1;
        
        if(n<=1)    return dp[n];
        
        return util(n);
    }
}

File: leet/70e_ClimbingStairs.py
class Solution:
    def climbStairs(self, n: int) -> int:
        if n in (0, 1, 2):
            return n

        pprev, prev = 1, 2
        
        for i in range(3, n+1):
            ans = prev + pprev
            pprev = prev
            prev = ans

        return ans
